File: JAVA/Irrgarten_JAVA/src/irrgarten/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.Random;

/**
 *
 * @author alejandro
 */
public class Dice {
    private static final int MAX_USES=5;
        private static final float MAX_INTELLIGENCE=10.0f;
        private static final float MAX_STRENGTH=10.0f;
        private static final float RESURRECT_PROB=0.3f;
        private static final int WEAPONS_REWARD=2;
        private static final int SHIELDS_REWARD=3;
        private static final int HEALTH_REWARD=5;
        private static final int MAX_ATTACK=3;
        private static final int MAX_SHIELD=2;
        
        
        private static final Random generator = new Random();
        
        public static int randomPos(int max){
            return generator.nextInt(max);
        }
        
        public static int whoStarts(int nplayers){
            return generator.nextInt(nplayers);
        }
        
        public static float randomIntelligence(){
            return generator.nextFloat()*MAX_INTELLIGENCE;
        }
        
        public static float randomStrength(){
            return generator.nextFloat()*MAX_STRENGTH;
        }
        
        public static boolean resurrectPlayer(){
            return generator.nextFloat() < RESURRECT_PROB;
        }
        
        public static int weaponsReward(){
            return generator.nextInt(WEAPONS_REWARD + 1);
        }
        
        public static int shieldsReward(){
            return generator.nextInt(SHIELDS_REWARD + 1);
        }
        
        public static int healthReward(){
            return generator.nextInt(HEALTH_REWARD + 1);
        }
        
        public static float weaponPower(){
            return generator.nextFloat()*MAX_ATTACK;
        }
        
        public static float shieldPower(){
            return generator.nextFloat()*MAX_SHIELD;
        }
        
        public static int usesLeft(){
            return generator.nextInt(MAX_USES + 1);
        }
        
        public static float intensity(float competence){
            return generator.nextFloat()*competence;
        }
        
        public static boolean discardElement(int usesLeft){
            if(usesLeft >= MAX_USES){
                return false;
            }else if(usesLeft == 0){
                return true;
            }else{
                float probability;
                probability = (float)(MAX_USES - usesLeft) / MAX_USES;
                return generator.nextFloat() < probability;
            }
        }
        
    }

File: RUBY/PRACTICA1/Practica-1/irrgarten_1/dice.rb
# frozen_string_literal: true
module Irrgarten
  class Dice
    @@MAX_USES = 5
    @@MAX_INTELLIGENCE = 10.0
    @@MAX_STRENGTH = 10.0
    @@RESURRECT_PROB = 0.3
    @@MAX_RESURRECT_PROB = 1.0
    @@WEAPONS_REWARD = 2
    @@SHIELDS_REWARD = 3
    @@HEALTH_REWARD = 5
    @@MAX_ATTACK = 3
    @@MAX_SHIELD = 2

    @@generator = Random.new

    class << self
      attr_reader :MAX_USES, :MAX_INTELLIGENCE, :MAX_STRENGTH
      attr_reader :RESURRECT_PROB, :WEAPONS_REWARD, :SHIELDS_REWARD, :HEALTH_REWARD
      attr_reader :MAX_ATTACK, :MAX_SHIELD

      def roll(sides=6)
        @@generator.rand(1..sides)
      end
    end

    def self.random_pos(max)
      raise ArgumentError, "EL valor tiene que ser mayor a 0" if max <= 0

      @@generator.rand(0...max)
    end

    def self.who_starts(nplayers)
      raise ArgumentError, "El numero de jugadores tiene que ser mayor a 0" if nplayers <= 0

      @@generator.rand(0...nplayers)
    end

    def self.random_intelligente
      @@generator.rand(0...MAX_INTELLIGENCE)
    end

    def self.random_strength
      @@generator.rand(0...MAX_STRENGTH)
    end

    def self.resurrect_player
      @@generator.rand(0...MAX_RESURRECT_PROB) < RESURRECT_PROB
    end

    def self.weapons_reward
      @@generator.rand(0...WEAPONS_REWARD)
    end

    def self.shields_reward
      @@generator.rand(0...SHIELDS_REWARD)
    end

    def self.health_reward
      @@generator.rand(0...HEALTH_REWARD)
    end

    def self.weapon_power
      @@generator.rand(0...MAX_ATTACK)
    end

    def self.shield_power
      @@generator.rand(0..MAX_SHIELD)
    end

    def self.uses_left
      @@generator.rand(0..MAX_USES)
    end

    def self.intensity(competence)
      @@generator.rand(0...competence)
    end

    def self.discard_element(uses_left)
      if @@generator.rand(1..@@MAX_USES) > uses_left
        return true
      end
      false
    end
  end
end